File: fushi/test/pages/popup_pending_scroll_restore_test.js
// 弹窗内原地跳转：后退 / 前进回到历史页时的滚动位恢复（`__fushiApplyPendingScrollTop`）。
//
// Dart 在 renderPopup() 之前把该页离开时的 scrollTop 写进 window.__fushiPendingScrollTop；
// 内容分批进 DOM，所以 popup.js 在三处应用它：首发 popupRendered（仍在渲染，够高才滚）、
// 每个尾批切片后（够高才滚）、尾批完成（final，兜底必滚、浏览器自行夹紧）。
//
// 用 Node 真执行 popup.js（vm + 极简假 DOM，见 _popup_dom_host.js），直接驱动这个函数：
//   1. pending = 0 / 负数 → 什么都不动。
//   2. 非 final 且文档还不够高 → 不滚、pending 保留（等下一片）。
//   3. 非 final 且够高 → 滚到位、pending 清零（同一份 pending 不会影响下一次渲染）。
//   4. final 且不够高 → 照样写 scrollTop（浏览器夹紧）、pending 清零。
//
// Run: node fushi/test/pages/popup_pending_scroll_restore_test.js
// (also driven from popup_pending_scroll_restore_test.dart inside `flutter test`).

const assert = require('assert');
const vm = require('vm');
const { loadPopup } = require('./_popup_dom_host.js');

function makeScroller(scrollHeight, clientHeight) {
  return { scrollHeight, clientHeight, scrollTop: 0 };
}

function run(pending, scroller, isFinal) {
  const sandbox = loadPopup();
  sandbox.document.scrollingElement = scroller;
  sandbox.window.__fushiPendingScrollTop = pending;
  vm.runInContext(`__fushiApplyPendingScrollTop(${isFinal ? 'true' : 'false'})`, sandbox);
  return { scrollTop: scroller.scrollTop, pending: sandbox.window.__fushiPendingScrollTop };
}

// 1. 没有待恢复位：不动。
{
  const r = run(0, makeScroller(2000, 400), true);
  assert.strictEqual(r.scrollTop, 0);
  assert.strictEqual(r.pending, 0);
  const neg = run(-5, makeScroller(2000, 400), true);
  assert.strictEqual(neg.scrollTop, 0);
}

// 2. 非 final、还不够高：不滚，pending 保留给下一片。
{
  const r = run(1200, makeScroller(900, 400), false);
  assert.strictEqual(r.scrollTop, 0, '文档不够高时提前滚会被夹到底、随后又被后续内容顶回去');
  assert.strictEqual(r.pending, 1200, 'pending 必须保留，等下一个切片再试');
}

// 3. 非 final、够高：滚到位并清零。
{
  const r = run(1200, makeScroller(2000, 400), false);
  assert.strictEqual(r.scrollTop, 1200);
  assert.strictEqual(r.pending, 0, '应用后清零：下一次渲染不得再被这份 pending 影响');
}

// 4. final 兜底：不够高也写（浏览器夹紧），并清零。
{
  const r = run(1200, makeScroller(900, 400), true);
  assert.strictEqual(r.scrollTop, 1200, 'final 一律应用（真实浏览器会夹到 scrollHeight-clientHeight）');
  assert.strictEqual(r.pending, 0);
}

// 5. 装载后初值：popup.js 自己把 pending 初始化为 0（热槽跨查词不带脏值）。
{
  const sandbox = loadPopup();
  assert.strictEqual(sandbox.window.__fushiPendingScrollTop, 0);
}

console.log('popup_pending_scroll_restore: all assertions passed');

File: fushi/test/pages/popup_pending_scroll_restore_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// 弹窗内原地跳转：后退 / 前进回到历史页时的滚动位恢复（Node 真执行 popup.js，
/// 见同名 .js）。
///
/// 守的是 `__fushiApplyPendingScrollTop` 的可观测行为：没有 pending 不动；非 final
/// 且文档不够高时**不滚、pending 保留**（提前滚会被夹到底、再被后续内容顶回去）；
/// 够高即滚到位并清零；final 兜底一律应用并清零。Dart 侧 / 三处调用点的接线守卫在
/// dictionary_popup_inplace_navigation_test.dart。无 node 时 skip（与
/// popup_render_tail_batching_test.dart 同款约定）。
void main() {
  test('popup pending scroll restore (executes popup.js via node)', () async {
    final String? nodeExe = _resolveNode();
    if (nodeExe == null) {
      markTestSkipped('node not found on PATH; skipping JS behavior execution');
      return;
    }

    final File jsTest = File('test/pages/popup_pending_scroll_restore_test.js');
    expect(jsTest.existsSync(), isTrue,
        reason: 'behavior harness ${jsTest.path} must exist');

    final ProcessResult result = await Process.run(
      nodeExe,
      <String>[jsTest.path],
      workingDirectory: Directory.current.path,
    );

    expect(
      result.exitCode,
      0,
      reason: 'popup pending-scroll JS behavior test failed.\n'
          'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
    );
    expect(result.stdout.toString(), contains('all assertions passed'));
  });
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) return name;
    } on ProcessException {
      // try next candidate
    }
  }
  return null;
}
